### ISSUE TITLE
RFC: Better CSS composability

### ISSUE THREAD

**Issue by madou** [atlassian-labs/compiled-css-in-js#17]
Currently you can compose CSS if:

1. It is variable is statically analyzable to be CSS like
1. Accept that the style will be inlined into the compiled component (multiple uses of composing = multiple declarations - gzip would make this w/e I suppose but it's still more code to execute!)

And you can do it via module boundaries if:

1. Typechecking with TypeScript is turned on (this means it doesn't work if it's off, and it doesn't work if you're compiling with Babel, issue to make it better here https://github.com/atlassian-labs/compiled-css-in-js/issues/66)

---

Use case is to transform variable declarations and then re-use them as variables instead of always inlining them into the `style` tags. I want to lean on the name `mixin` over `css` because this does not, and will never, return a className. `css` is overloaded in other css-in-js libraries to return a form of className.

Transforms:

```jsx
const borderColor = 'black';

const myMixin = (props) => mixin({
  border: `1px solid ${borderColor}`,
  color: props.color,
});

// mixin return type is the input object
```

To:

```jsx
const myMixin = (props) => ({
  variables: { '--color-a': props.color },
  css: `
    border: 1px solid black; /* variable isnt in the closure so it gets compiled in */
    color: var(--color-a); /* variable was in closure so it gets xformed to a css var */
  `
});
```

...which then can be reused in components

From:

```jsx
import { myMixin } from './mixins';

const MyDiv = (props) => (
  <div css={{ '&': myMixin({ color: props.color })  }}>
    hello world
  </div>
);
```

To:

```jsx
import { myMixin } from './mixins';

const MyDiv = (props) => {
  const mixin = myMixin({ color: props.color });

  return (
    <>
      <style>{`.a { ${mixin.css} }`}</style>
      <div style={mixin.variables} className="a">hello world</div>
    </>
);
```

**Comment by madou** [comment]
This is basically the keyframes api #14 - made them in tandem

**Comment by madou** [comment]
Made a comment about this in https://github.com/atlassian-labs/compiled-css-in-js/issues/14#issuecomment-620430297

tldr dunno if this is as good idea.

**Comment by madou** [comment]
It's not. Closing.
